text/0000-variadic-tuples.md

```diff
@@ -0,0 +1,1006 @@
+- Implementations will be easier to read and maintain
+- The compiler will compile implementation only for required tuple arity
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Let's call a _variadic tuple type_ a tuple type with an arbitrary arity and a _variadic tuple_ an instance of a variadic tuple type.
+
+A variadic tuple type is declared with `(..T)` and a variadic tuple type can be expanded with `(..Vec<T>)`.
+
+Note: To illustrate the RFC, we will use the current implementation of the `Hash` trait for tuples.
+
+```rust
+// Quote from Rust source code
+// This macro implements `Hash` for a tuple.
+// It is used like this: `impl_hash_tuple! { A B C D E F }` for a 6-arity tuple.
+macro_rules! impl_hash_tuple {
+    () => (
+        #[stable(feature = "rust1", since = "1.0.0")]
+        impl Hash for () {
+            fn hash<H: Hasher>(&self, _state: &mut H) {}
+        }
+    );
+
+    ( $($name:ident)+) => (
+        #[stable(feature = "rust1", since = "1.0.0")]
+        impl<$($name: Hash),+> Hash for ($($name,)+) where last_type!($($name,)+): ?Sized {
+            #[allow(non_snake_case)]
+            fn hash<S: Hasher>(&self, state: &mut S) {
+                let ($(ref $name,)+) = *self;
+                $($name.hash(state);)+
+            }
+        }
+    );
+}
+
+macro_rules! last_type {
+    ($a:ident,) => { $a };
+    ($a:ident, $($rest_a:ident,)+) => { last_type!($($rest_a,)+) };
+}
+```
+
+## Variadic tuple type
+
+### Declaration
+
+Variadic tuple types are always declared in a generic parameter group.
+
+There are two different syntaxes:
+
+1. `(..T)`: declare a single variadic tuple type identified by `T`
+2. `(..(T1, T2, ..., Tn))`: declare n variadic tuple types identified by `T1`, `T2`, ..., `Tn`, all these variadic tuple types have the same arity.
+
+Declaration examples:
+
+- `struct VariadicStruct<(..T1)>` : declares a struct with a variadic tuple type identified by `T1` in its generic parameters
+- `fn my_func<(..T1), (..T2)>()`: a function can have variadic tuple type parameters
+- `impl<(..Head)>`: is an implementation block that uses a variadic tuple type identified by `Head`
+- `impl<A, B, C, (.._Tail)>`:  same as above, but with other generic parameters
+- `impl<A, B, (..C), (..D)>`: there can be several variadic tuple types declared in a generic parameter group
+- `impl<A, B, (..(C, D)), (..E)>`: and we can mix both syntaxes
+
+Usage examples:
+
+```rust
+struct VariadicStruct<(..T)>
+VariadicStruct<(usize,)>                // => (..T) matches (usize,)
+VariadicStruct<(usize, bool)>   // => (..T) matches (usize, bool)
+
+impl <(..(T1, T2))> VariadicStruct<(..(T1, T2))> { ... }
+VariadicStruct::<((usize, bool),)>
+// (..(T1, T2)) matches ((usize, bool),)
+// (..T1) is (usize,)
+// (..T2) is (bool,)
+VariadicStruct::<((usize, bool), (String, i8))> // (..(T1, T2)) matches ((usize, bool), (String, i8))
+// (..T1) is (usize, String)
+// (..T2) is (bool, i8)
+```
```

Comment:
Again, don't go with full code; just focus on the tuples. So, say that `(..T)` would match `()`, `(u32,)`, etc. whereas `(Head, ..Tail)` wouldn't match `()` and `(First, ..Middle, Last)` wouldn't match `(u32,)`.

Reply:
I am not sure that I agree; the code here fairly effectively communicates concepts that are difficult to explain through prose.  (also, as per my other comment, `(Head, ..Tail)` is not allowed in this position)